src/courses/tests/test_views.py

```diff
@@ -664,6 +666,29 @@ def test_SemesterUpdate__delete_data__view(self):
 
         self.assertFalse(ExcludedDate.objects.exists())
 
+    @patch('profile_manager.models.Profile.xp_per_course')
+    def test_SemesterDCloses__student_with_negative_xp__view(self, xp_per_course):
+        xp_per_course.return_value = -10
+        self.client.force_login(self.test_teacher)
+
+        post_data = {
+            'first_day': '2020-10-15', 'last_day': '2020-12-15',
+            **generate_formset_data(ExcludedDateFormset, quantity=0)
+        }
+        response = self.client.post(reverse('courses:semester_create'), data=post_data)
+        self.assertRedirects(response, reverse('courses:semester_list'))
+        self.assertEqual(Semester.objects.count(), 2)
+
+        semester = Semester.objects.first()
+        student = baker.make(User)
+        course = baker.make(Course)
+        baker.make(CourseStudent, user=student, course=course, semester=semester)
+
+        response = self.client.post(reverse('courses:end_active_semester'))
+        self.assertWarningMessage(response)
+        message = self.get_message_list(response)[0]
+        self.assertIn('There are some students with negative XP', str(message))
```

Comment:
Looking good.  Does this test check if we land on a good page somewhere?  i.e. eventually we get a 200 response?

Reply:
added